### ISSUE TITLE
Injecting/wrapping globals for non-global module

### ISSUE THREAD

**Issue by sberney** [systemjs/systemjs#1003]
Thanks for your help, I can tell you have a lot to do.

It's my impression that the `globals` meta operates only for cjs and global modules (a search turned up #977 but that's not the place I remember reading about this). I wonder if I misunderstand the purpose of `globals`, because I don't see why it should not apply to amd.

One application I am loading is compiled as a bunch of AMD modules, but it expects jquery to be a global. I can't just wrap like I would with requirejs below (`TypeError: AMD module http://localhost/js/app3/foo.js did not define`):

```javascript
// app3/foo.js
require(['jquery'], function ($) {
  define('foo', [], function () {
    $.post( ... );
  });
});
```

Even if I could, I would wonder why there isn't something that does this automatically for AMD modules; where by this, I mean ensuring that a library is loaded and bound to an arbitrary symbol during evaluation, along the lines of

```javascript
'meta': {
  'js/app3/*': {
    'globals': {
      '$123': 'jquery'
    }
  }
}
```

I'm had a similar problem with lodash. I want to allow `_` to be accessible when lodash templates are evaluating, but keep `_` out of the global namespace. However, there I was able to use the `instantiate` hook. I don't want to generally be playing with instantiate...

Related aside: A year ago you said that `instantiate` needs to return an object in the form `{ deps: [], execute: function () { } }`, [here](https://github.com/systemjs/builder/issues/22#issuecomment-69720450), but [this line](https://github.com/systemjs/systemjs/blob/master/lib/plugins.js#L226) uses whatever instantiate returns as the actual module value (it doesn't seem to understand an object of the aforementioned form). If loader instantiates are supposed to return the object of interest itself, I haven't found the documentation for it (but it seems to be the thing to do).

Is there a simple way to make sure jquery (and a few others) are loaded as the symbols that are expected by AMD modules defined in a single file? I'll keep looking, but just adding them to `deps` doesn't seem to work, unless of course I let the dependencies live perpetually on window.

**Comment by guybedford** [comment]
Sure, thanks for understanding. That said, shorter messages would be preferable :)

We only do the globals feature for CommonJS and global modules because those are the module formats that get their source rewritten in SystemJS. AMD can be loaded by direct <script> injection so doesn't get any rewriting - it is the final source text.

Rather I'd suggest managing this through the use of `metadata.deps` on jQuery, and then using a version of jQuery that does define a global.

**Comment by sberney** [comment]
Thanks! That makes sense. It would be a nice feature to have. I'll probably end up implementing something myself.

**Comment by sberney** [comment]
I did a [proof of concept](https://github.com/sberney/systemjs/commit/b92096cabb35189d3bc1c1556bbd5dfc81b57b62) for this feature (there's a debugger or two in there), but after discussion with team members, it does not seem worth it to maintain a custom distribution of systemjs. I see what you mean when you say that AMD is not rewritten -- the main issue it seems is that the source is available for the javascript page and not for each module, currently.

This implementation ~~sucks~~ breaks the trace: I've turned the native source code into a string and back during the execution stage of module loading  ... A better implementation would apply inside of the AMD extension `instantiate` hook, apply regular expressions to locate factories, inserting the appropriate symbols there. Adding dependency names could be done this way, or in a way more similar to the approach here. Also, it may not be perfect, as in one AMD file I tested this against, there is

```javascript
var foobar = 'hello';
define('something', [], function () {
  return foobar;
});
```

which in my version complains that `foobar` doesn't exist. But I thought I would share my work on this, in case one day in the future you consider the `globals` for AMD modules further.

I didn't implement matching against the meta dictionary ... with this simple implementation, anything using that edition of define gets the dependencies injected. That is undesirable, because the `globals` attribute is defined in a `meta`, and should thus only apply to that meta (multiple AMD definitions make this distinction necessary).

I also didn't consider cyclical dependency resolution. I don't know that much about it.
